File: Python/Remove Duplicates from Sorted Array II.py
# Follow up for "Remove Duplicates":
# What if duplicates are allowed at most twice?

# For example,
# Given sorted array A = [1,1,1,2,2,3],

# Your function should return length = 5, and A is now [1,1,2,2,3]


class Solution:
    # @param A a list of integers
    # @return an integer
    def removeDuplicates(self, A):
		i = 1
		leng = len(A)
		if leng == 0:
		       return 0
		current = A[0]
		currentIndex = 0
		currentCount = 0
		while i < len(A):
			if current == A[i]:
				if currentCount == 0:
					currentCount = 1
					current = A[i]
					A[currentIndex + 1] = A[i]
					currentIndex = currentIndex + 1
					i = i + 1
				else:
					leng = leng - 1
					i = i + 1	
			else:
				currentCount = 0
				current = A[i]
				A[currentIndex + 1] = A[i]
				currentIndex = currentIndex + 1
				i = i + 1
		A = A[0:leng]
		return leng
		

File: Java/Remove Duplicates from Sorted Array II.java
class Remove Duplicates from Sorted Array II{
	public int removeDuplicates(int[] A) {
        if(A == null || A.length == 0)
        	return 0;
        int len = A.length;
        int twice = 0;
        int pre = 0;
        int curr = 1;
        while(curr < A.length){
        	if(A[curr] != A[pre]){
        		A[pre+1] = A[curr];
        		curr++;
        		pre++;
        		twice = 0;
        	}
        	else if(A[curr] == A[pre] && twice == 0){
        		twice = 1;
        		A[pre+1] = A[curr];
        		pre++;
        		curr++;
        	}
        	else{
        		len--;
        		curr++;
        	}
        }
        int[] newA = Arrays.copyOfRange(A,0,len);
        return len;
    }
}